Index: wireshark/epan/dissectors/packet-rsl.c
===================================================================
--- wireshark.orig/epan/dissectors/packet-rsl.c
+++ wireshark/epan/dissectors/packet-rsl.c
@@ -2291,7 +2291,7 @@
 
 	proto_tree_add_text(ie_tree, tvb,offset,length,"Layer 3 message");
 	next_tvb = tvb_new_subset(tvb, offset, length, length);
-	/* call_dissector(gsm_a_dtap_handle, next_tvb, pinfo, top_tree);*/
+	call_dissector(gsm_a_ccch_handle, next_tvb, pinfo, top_tree);
 
 	offset = offset + length;
 
